File: src/epub_optimizer/repair.py
from __future__ import annotations

import mimetypes
import posixpath
import re
from pathlib import Path, PurePosixPath
from urllib.parse import unquote, urlsplit

from lxml import etree

from epub_optimizer.epubcheck import EpubCheckFinding

OPF_NS = "http://www.idpf.org/2007/opf"


def repair_workspace(
    work_dir: Path,
    package_path: str,
    package_tree: etree._ElementTree,
    findings: list[EpubCheckFinding],
) -> list[str]:
    """Apply only deterministic, loss-minimizing EPUB repairs in-place."""
    codes = {finding.code.upper() for finding in findings}
    broken_reference_codes = {"RSC-007", "RSC-012"}
    repair_broken_references = bool(codes & broken_reference_codes)
    repair_missing_manifest_files = "OPF-003" in codes
    repair_missing_manifest_entries = "OPF-012" in codes
    metadata_findings = [
        finding
        for finding in findings
        if finding.path and _normalized_package_path(unquote(finding.path))
        == _normalized_package_path(package_path)
    ]
    repair_metadata = _repair_opf_metadata(package_tree, metadata_findings)
    affected_paths = {
        _normalized_package_path(unquote(finding.path))
        for finding in findings
        if finding.code.upper() in broken_reference_codes and finding.path
    }
    root = package_tree.getroot()
    manifest = root.find(f"{{{OPF_NS}}}manifest")
    if manifest is None:
        manifest = root.find("manifest")
    if manifest is None:
        return repair_metadata
    package_dir = posixpath.dirname(package_path)
    actions: list[str] = list(repair_metadata)
    items = [e for e in manifest if isinstance(e.tag, str) and etree.QName(e).localname == "item"]
    ncx_paths = {
        _normalized_package_path(unquote(finding.path))
        for finding in findings
        if finding.code.upper() == "NCX-001" and finding.path
    }
    if ("OPF-030" in codes and repair_metadata) or ncx_paths:
        actions.extend(
            _synchronize_ncx_uid(
                work_dir,
                package_dir,
                package_tree,
                items,
                ncx_paths=ncx_paths or None,
            )
        )
    if not (
        repair_broken_references
        or repair_missing_manifest_files
        or repair_missing_manifest_entries
    ):
        return actions
    ids = {e.attrib.get("id", "") for e in items}
    missing_items = []
    spine_ids: set[str] = set()
    spine = root.find(f"{{{OPF_NS}}}spine")
    if spine is None:
        spine = root.find("spine")
    if spine is not None:
        spine_ids = {ref.attrib.get("idref", "") for ref in spine}
    for item in items:
        href = item.attrib.get("href")
        if not href:
            continue
        target = _safe_resolve(work_dir, package_dir, href)
        if target is None:
            continue
        if (
            repair_missing_manifest_files
            and not target.is_file()
            and item.attrib.get("id", "") not in spine_ids
        ):
            missing_items.append(item)
    if missing_items:
        missing_ids = {i.attrib.get("id", "") for i in missing_items}
        for item in missing_items:
            href = item.attrib.get("href", "")
            manifest.remove(item)
            actions.append(f"Removed missing manifest item: {href}")
        spine = root.find(f"{{{OPF_NS}}}spine")
        if spine is None:
            spine = root.find("spine")
        if spine is not None:
            for ref in list(spine):
                if ref.attrib.get("idref") in missing_ids:
                    spine.remove(ref)
                    actions.append(f"Removed stale spine reference: {ref.attrib.get('idref', '')}")
        items = [
            e for e in manifest if isinstance(e.tag, str) and etree.QName(e).localname == "item"
        ]
        ids = {e.attrib.get("id", "") for e in items}

    content_items = [
        i
        for i in items
        if i.attrib.get("media-type", "").lower()
        in {"application/xhtml+xml", "text/html", "application/x-dtbook+xml"}
    ]
    for item in content_items:
        href = item.attrib.get("href")
        if not href:
            continue
        rel_path = posixpath.normpath(posixpath.join(package_dir, unquote(href)))
        repair_document_references = (
            repair_broken_references
            and bool(affected_paths)
            and _normalized_package_path(rel_path) in affected_paths
        )
        if not repair_document_references and not repair_missing_manifest_entries:
            continue
        content_file = _safe_resolve(work_dir, "", rel_path)
        if content_file is None:
            continue
        if not content_file.is_file():
            continue
        try:
            tree = etree.parse(
                str(content_file),
                etree.XMLParser(resolve_entities=False, no_network=True, recover=True),
            )
        except (OSError, etree.XMLSyntaxError):
            continue
        changed = False
        elements = tree.getroot().xpath("//*[@href or @src or @data or @*[local-name()='href']]")
        for element in list(elements):
            attrs = [a for a in ("href", "src", "data") if a in element.attrib]
            attrs.extend(
                a
                for a in element.attrib
                if etree.QName(a).localname == "href" and a not in attrs
            )
            for attr in attrs:
                value = element.attrib.get(attr)
                if not value:
                    continue
                parsed = urlsplit(value)
                if parsed.scheme or parsed.netloc:
                    continue
                if not parsed.path and parsed.fragment:
                    if not repair_document_references:
                        continue
                    if _fragment_exists(content_file, unquote(parsed.fragment)):
                        continue
                    _remove_reference(element, attr, actions, value)
                    changed = True
                    continue
                if not parsed.path:
                    continue
                target_rel = posixpath.normpath(
                    posixpath.join(posixpath.dirname(rel_path), unquote(parsed.path))
                )
                if target_rel.startswith("../") or target_rel.startswith("/"):
                    continue
                target = _safe_resolve(work_dir, "", target_rel)
                if target is None:
                    continue
                if parsed.fragment and (not target.is_file() or _is_html(target_rel)):
                    if target.is_file() and _fragment_exists(target, unquote(parsed.fragment)):
                        continue
                    if not target.is_file() and parsed.path:
                        if not repair_document_references:
                            continue
                        _remove_reference(element, attr, actions, value)
                        changed = True
                    else:
                        element.attrib[attr] = parsed.path or ""
                        actions.append(f"Removed broken fragment: {value}")
                        changed = True
                    continue
                if not target.is_file():
                    if not repair_document_references:
                        continue
                    _remove_reference(element, attr, actions, value)
                    changed = True
                    continue
                if repair_missing_manifest_entries and not _manifest_contains(
                    items, target_rel, package_dir
                ):
                    media = mimetypes.guess_type(target.name)[0]
                    if media:
                        new_id = _unique_id(ids, target.stem or "resource")
                        new_href = posixpath.relpath(target_rel, package_dir or ".")
                        new_item = etree.Element(
                            _qualified_like(manifest, "item"),
                            attrib={"id": new_id, "href": new_href, "media-type": media},
                        )
                        manifest.append(new_item)
                        items.append(new_item)
                        ids.add(new_id)
                        actions.append(f"Added manifest item for referenced resource: {new_href}")
                        changed = True

        if changed:
            tree.write(
                str(content_file), encoding="utf-8", xml_declaration=True, pretty_print=False
            )
    return actions


def _synchronize_ncx_uid(
    work_dir: Path,
    package_dir: str,
    package_tree: etree._ElementTree,
    items: list[etree._Element],
    *,
    ncx_paths: set[str] | None,
) -> list[str]:
    root = package_tree.getroot()
    identifier_id = root.attrib.get("unique-identifier", "")
    identifiers = root.xpath(
        "//*[local-name()='metadata']/*[local-name()='identifier' and @id=$identifier_id]",
        identifier_id=identifier_id,
    )
    if not identifiers or not (identifiers[0].text or "").strip():
        return []
    identifier = (identifiers[0].text or "").strip()
    actions = []
    for item in items:
        if item.attrib.get("media-type", "").lower() != "application/x-dtbncx+xml":
            continue
        href = item.attrib.get("href", "")
        decoded_href = unquote(href)
        package_path = _normalized_package_path(posixpath.join(package_dir, decoded_href))
        if ncx_paths is not None and package_path not in ncx_paths:
            continue
        ncx_path = _safe_resolve(work_dir, package_dir, decoded_href)
        if ncx_path is None or not ncx_path.is_file():
            continue
        try:
            tree = etree.parse(
                str(ncx_path),
                etree.XMLParser(resolve_entities=False, no_network=True, recover=False),
            )
        except (OSError, etree.XMLSyntaxError):
            continue
        uid_nodes = tree.getroot().xpath(
            "//*[local-name()='meta' and translate(@name, 'ABCDEFGHIJKLMNOPQRSTUVWXYZ', "
            "'abcdefghijklmnopqrstuvwxyz')='dtb:uid']"
        )
        if not uid_nodes or uid_nodes[0].attrib.get("content") == identifier:
            continue
        uid_nodes[0].attrib["content"] = identifier
        tree.write(str(ncx_path), encoding="utf-8", xml_declaration=True, pretty_print=False)
        actions.append(f"Synchronized NCX identifier with OPF identifier: {href}")
    return actions


def _repair_opf_metadata(
    package_tree: etree._ElementTree, findings: list[EpubCheckFinding]
) -> list[str]:
    """Conservative, finding-gated repairs for common OPF metadata errors."""
    actions: list[str] = []
    root = package_tree.getroot()
    codes = {f.code.upper() for f in findings}
    metadata = root.find(f"{{{OPF_NS}}}metadata")
    if metadata is None:
        metadata = root.find("metadata")
    if metadata is None:
        return actions
    if "OPF-030" in codes:
        identifiers = [
            e
            for e in metadata
            if etree.QName(e).localname == "identifier" and (e.text or "").strip()
        ]
        if identifiers:
            chosen = next((e for e in identifiers if e.attrib.get("id")), identifiers[0])
            ident = chosen.attrib.get("id")
            if not ident:
                ident = "bookid"
                used = {e.attrib.get("id") for e in metadata if e.attrib.get("id")}
                i = 2
                while ident in used:
                    ident = f"bookid-{i}"
                    i += 1
                chosen.set("id", ident)
                actions.append(f"Assigned stable id to dc:identifier: {ident}")
            root.set("unique-identifier", ident)
            actions.append(f"Pointed package unique-identifier to dc:identifier: {ident}")
    if "OPF-028" in codes:
        prefix_attr = root.attrib.get("prefix", "")
        declared_prefixes = {
            token.rstrip(":") for token in prefix_attr.split()[::2] if token.endswith(":")
        }
        # Calibre's known vocabulary can be declared losslessly. Unknown
        # vendor metadata is preserved and remains a reported legacy issue.
        for finding in findings:
            if finding.code.upper() != "OPF-028":
                continue
            match = re.search(r'prefix:\s*"([A-Za-z][\w-]*)"', finding.message, re.I)
            if not match:
                continue
            prefix = match.group(1)
            if prefix.lower() == "calibre" and prefix not in declared_prefixes:
                declaration = "calibre: https://calibre-ebook.com"
                root.set("prefix", f"{prefix_attr} {declaration}".strip())
                prefix_attr = root.attrib["prefix"]
                declared_prefixes.add(prefix)
                actions.append("Declared known calibre metadata vocabulary")
                continue
    if "RSC-005" in codes:
        invalid_properties = {
            match.group(1)
            for finding in findings
            for match in [re.search(r'Property "([^"]+)" must refine', finding.message)]
            if match
        }
        valid_refinement_targets = {
            "role": {"creator", "contributor", "publisher"},
            "file-as": {"creator", "contributor"},
            "title-type": {"title"},
        }
        for elem in list(metadata):
            if etree.QName(elem).localname != "meta" or "refines" not in elem.attrib:
                continue
            # Only discard malformed refinement metadata; primary dc:* remains intact.
            ref = elem.attrib.get("refines", "")
            prop = elem.attrib.get("property", "")
            target = next((e for e in metadata if e.attrib.get("id") == ref[1:]), None)
            target_is_invalid = False
            allowed_targets = valid_refinement_targets.get(prop)
            if prop in invalid_properties and target is not None and allowed_targets is not None:
                target_is_invalid = etree.QName(target).localname not in allowed_targets
            if (
                not ref.startswith("#")
                or not prop
                or target is None
                or target_is_invalid
            ):
                metadata.remove(elem)
                actions.append("Removed invalid metadata refinement")
    return actions


def _normalized_package_path(path: str) -> str:
    return posixpath.normpath(path.replace("\\", "/").lstrip("/"))


def _remove_reference(element, attr, actions, value):
    local = etree.QName(element).localname.lower()
    if local in {"a", "area"} and attr == "href":
        element.attrib.pop(attr, None)
        actions.append(f"Removed broken link target (text preserved): {value}")
    elif local == "img":
        parent = element.getparent()
        if parent is not None:
            alt = element.attrib.get("alt", "").strip()
            if alt:
                replacement = etree.Element(_qualified_like(element, "span"))
                replacement.text = alt
                replacement.tail = element.tail
                parent.replace(element, replacement)
                actions.append(f"Replaced broken image with alt text: {value}")
            else:
                parent.remove(element)
                actions.append(f"Removed broken image without alt text: {value}")
    elif local in {"object", "audio", "video"}:
        element.attrib.pop(attr, None)
        actions.append(f"Removed broken resource attribute; fallback preserved: {value}")
    elif local in {"link", "script", "image", "source"}:
        parent = element.getparent()
        if parent is not None:
            parent.remove(element)
            actions.append(f"Removed broken resource element: {value}")
    else:
        element.attrib.pop(attr, None)
        actions.append(f"Removed broken reference: {value}")


def _resolve(work_dir: Path, package_dir: str, href: str) -> Path:
    raw = posixpath.join(package_dir, href)
    if raw.startswith("/") or PurePosixPath(href).is_absolute():
        raise ValueError("absolute EPUB path")
    candidate = (work_dir / Path(*PurePosixPath(posixpath.normpath(raw)).parts)).resolve()
    root = work_dir.resolve()
    if candidate != root and root not in candidate.parents:
        raise ValueError("EPUB path escapes workspace")
    return candidate


def _safe_resolve(work_dir: Path, package_dir: str, href: str) -> Path | None:
    try:
        return _resolve(work_dir, package_dir, href)
    except ValueError:
        return None


def _is_html(path: str) -> bool:
    return PurePosixPath(path).suffix.lower() in {".xhtml", ".html", ".htm"}


def _fragment_exists(path: Path, fragment: str) -> bool:
    try:
        root = etree.parse(
            str(path), etree.XMLParser(resolve_entities=False, no_network=True, recover=True)
        ).getroot()
        return bool(root.xpath("//*[@id=$id or @name=$id]", id=fragment))
    except (OSError, etree.XMLSyntaxError):
        return False


def _manifest_contains(items, target_rel: str, package_dir: str) -> bool:
    target = posixpath.normpath(target_rel)
    for item in items:
        href = item.attrib.get("href", "")
        if posixpath.normpath(posixpath.join(package_dir, href)) == target:
            return True
    return False


def _unique_id(ids: set[str], stem: str) -> str:
    base = "optimizer-" + "".join(c if c.isalnum() else "-" for c in stem).strip("-") or "resource"
    candidate = base
    index = 2
    while candidate in ids:
        candidate = f"{base}-{index}"
        index += 1
    return candidate


def _qualified_like(element, local_name):
    namespace = etree.QName(element).namespace
    return f"{{{namespace}}}{local_name}" if namespace else local_name

File: src/repair.rs
use crate::xml::Node;
use crate::{
    archive,
    uri::{Uri, decode},
    xml::Xml,
};
use anyhow::Result;
use std::{
    collections::{BTreeMap, BTreeSet},
    path::Path,
};

pub fn is_content(item: &Node) -> bool {
    matches!(
        item.get("media-type").to_lowercase().as_str(),
        "application/xhtml+xml" | "text/html" | "application/x-dtbook+xml"
    ) && !item.get("href").is_empty()
}
pub fn html(path: &str) -> bool {
    matches!(
        archive::extension(Path::new(path)).as_str(),
        "html" | "htm" | "xhtml"
    )
}
pub fn anchors(root: &Node) -> BTreeSet<String> {
    root.all()
        .iter()
        .flat_map(|n| {
            ["id", "name"]
                .into_iter()
                .filter(|a| n.has(a))
                .map(|a| n.get(a))
        })
        .collect()
}
pub fn fragment_exists(path: &Path, fragment: &str) -> bool {
    Xml::read(path, true).is_ok_and(|xml| anchors(&xml.root).contains(fragment))
}

pub fn hyperlinks(work: &Path, package_dir: &str, items: &[Node]) -> Result<Vec<String>> {
    let mut documents = Vec::new();
    let mut by_path = BTreeMap::new();
    let mut by_name: BTreeMap<String, BTreeSet<String>> = BTreeMap::new();
    let mut by_anchor: BTreeMap<String, BTreeSet<String>> = BTreeMap::new();
    for item in items.iter().filter(|i| is_content(i)) {
        let uri = Uri::parse(&item.get("href"));
        if uri.external || uri.path.is_empty() {
            continue;
        }
        let Ok(path) = archive::join(package_dir, &decode(&uri.path)) else {
            continue;
        };
        if !work.join(&path).is_file() || by_path.contains_key(&path) {
            continue;
        }
        let xml = Xml::read(&work.join(&path), true)?;
        let ids = anchors(&xml.root);
        for id in &ids {
            by_anchor
                .entry(id.clone())
                .or_default()
                .insert(path.clone());
        }
        by_name
            .entry(archive::basename(&path).into())
            .or_default()
            .insert(path.clone());
        by_path.insert(path.clone(), ids);
        documents.push((path, xml));
    }
    let mut actions = Vec::new();
    for (source, xml) in documents {
        let mut changed = false;
        for n in xml.root.named("a").iter().filter(|n| n.has("href")) {
            let href = n.get("href");
            let uri = Uri::parse(&href);
            if uri.external {
                continue;
            }
            let path = decode(&uri.path);
            let fragment = decode(&uri.fragment);
            let target = if path.is_empty() {
                Some(source.clone())
            } else {
                archive::join(archive::dirname(&source), &path).ok()
            };
            if let Some(target) = &target
                && work.join(target).is_file()
            {
                if fragment.is_empty() || (!by_path.contains_key(target) && !html(target)) {
                    continue;
                }
                if by_path
                    .get(target)
                    .is_some_and(|ids| ids.contains(&fragment))
                    || (!by_path.contains_key(target)
                        && fragment_exists(&work.join(target), &fragment))
                {
                    continue;
                }
            }
            let mut candidates = by_name
                .get(archive::basename(&path))
                .cloned()
                .unwrap_or_default();
            if !fragment.is_empty() {
                let matching = by_anchor.get(&fragment).cloned().unwrap_or_default();
                let common: BTreeSet<_> = candidates.intersection(&matching).cloned().collect();
                candidates = if common.is_empty() { matching } else { common };
            }
            if candidates.len() == 1 {
                let replacement =
                    uri.with_path(&archive::relative(&source, candidates.first().unwrap()));
                n.set("href", &replacement);
                actions.push(format!(
                    "Repaired hyperlink in {source}: {href} -> {replacement}"
                ));
            } else {
                n.del("href");
                actions.push(format!(
                    "Removed broken hyperlink in {source} (text preserved): {href}"
                ));
            }
            changed = true;
        }
        if changed {
            xml.write(&work.join(source))?;
        }
    }
    Ok(actions)
}

fn normalized_path(path: &str) -> String {
    archive::join("", path.replace('\\', "/").trim_start_matches('/'))
        .unwrap_or_else(|_| path.into())
}

pub fn workspace(
    work: &Path,
    package_path: &str,
    package: &Node,
    findings: &[crate::epubcheck::Finding],
) -> Result<Vec<String>> {
    let codes: BTreeSet<_> = findings.iter().map(|f| f.code.to_uppercase()).collect();
    let references = codes.contains("RSC-007") || codes.contains("RSC-012");
    let missing_files = codes.contains("OPF-003");
    let missing_entries = codes.contains("OPF-012");
    let metadata_findings: Vec<_> = findings
        .iter()
        .filter(|f| {
            f.path
                .as_ref()
                .is_some_and(|p| normalized_path(&decode(p)) == normalized_path(package_path))
        })
        .cloned()
        .collect();
    let mut actions = repair_metadata(package, &metadata_findings);
    let Some(manifest) = package.child("manifest") else {
        return Ok(actions);
    };
    let dir = archive::dirname(package_path);
    let mut items: Vec<_> = manifest
        .children()
        .into_iter()
        .filter(|n| n.name() == "item")
        .collect();
    let ncx_paths: BTreeSet<_> = findings
        .iter()
        .filter(|f| f.code.to_uppercase() == "NCX-001")
        .filter_map(|f| f.path.as_ref())
        .map(|p| normalized_path(&decode(p)))
        .collect();
    if (codes.contains("OPF-030") && !actions.is_empty()) || !ncx_paths.is_empty() {
        let identifier = package
            .child("metadata")
            .and_then(|m| {
                m.children().into_iter().find(|n| {
                    n.name() == "identifier" && n.get("id") == package.get("unique-identifier")
                })
            })
            .map(|n| n.content().trim().to_string())
            .unwrap_or_default();
        if !identifier.is_empty() {
            for item in &items {
                if item.get("media-type").to_lowercase() != "application/x-dtbncx+xml" {
                    continue;
                }
                let href = item.get("href");
                let Ok(path) = archive::join(dir, &decode(&href)) else {
                    continue;
                };
                if !ncx_paths.is_empty() && !ncx_paths.contains(&normalized_path(&path)) {
                    continue;
                }
                let Ok(xml) = Xml::read(&work.join(path.clone()), false) else {
                    continue;
                };
                if let Some(n) = xml
                    .root
                    .named("meta")
                    .iter()
                    .find(|n| n.get("name").to_lowercase() == "dtb:uid")
                    && n.get("content") != identifier
                {
                    n.set("content", &identifier);
                    xml.write(&work.join(path))?;
                    actions.push(format!(
                        "Synchronized NCX identifier with OPF identifier: {href}"
                    ));
                }
            }
        }
    }
    if !references && !missing_files && !missing_entries {
        return Ok(actions);
    }
    let spine_ids: BTreeSet<_> = package
        .child("spine")
        .map(|s| s.children().iter().map(|n| n.get("idref")).collect())
        .unwrap_or_default();
    if missing_files {
        items.retain(|item| {
            let href = item.get("href");
            if !href.is_empty()
                && !spine_ids.contains(&item.get("id"))
                && archive::join(dir, &href).is_ok_and(|p| !work.join(p).is_file())
            {
                item.remove_with_tail();
                actions.push(format!("Removed missing manifest item: {href}"));
                false
            } else {
                true
            }
        });
    }
    let affected: BTreeSet<_> = findings
        .iter()
        .filter(|f| matches!(f.code.to_uppercase().as_str(), "RSC-007" | "RSC-012"))
        .filter_map(|f| f.path.as_ref())
        .map(|p| normalized_path(&decode(p)))
        .collect();
    let content: Vec<_> = items.iter().filter(|i| is_content(i)).cloned().collect();
    let mut ids: BTreeSet<_> = items.iter().map(|i| i.get("id")).collect();
    for item in content {
        let Ok(path) = archive::join(dir, &decode(&item.get("href"))) else {
            continue;
        };
        let fix_refs = references && affected.contains(&normalized_path(&path));
        if !fix_refs && !missing_entries {
            continue;
        }
        let Ok(xml) = Xml::read(&work.join(&path), true) else {
            continue;
        };
        let mut changed = false;
        for n in xml.root.all() {
            let attrs = n.attrs();
            let mut names: Vec<_> = ["href", "src", "data"]
                .into_iter()
                .filter(|a| attrs.contains_key(*a))
                .map(str::to_string)
                .collect();
            names.extend(
                attrs
                    .keys()
                    .filter(|a| a.contains(':') && a.rsplit(':').next() == Some("href"))
                    .cloned(),
            );
            for attr in names {
                let value = n.get(&attr);
                if value.is_empty() {
                    continue;
                }
                let uri = Uri::parse(&value);
                if uri.external {
                    continue;
                }
                if uri.path.is_empty() {
                    if !uri.fragment.is_empty()
                        && fix_refs
                        && !fragment_exists(&work.join(&path), &decode(&uri.fragment))
                    {
                        remove_reference(&n, &attr, &value, &mut actions);
                        changed = true;
                    }
                    continue;
                }
                let Ok(target) = archive::join(archive::dirname(&path), &decode(&uri.path)) else {
                    continue;
                };
                let exists = work.join(&target).is_file();
                if !uri.fragment.is_empty() && (!exists || html(&target)) {
                    if exists && fragment_exists(&work.join(&target), &decode(&uri.fragment)) {
                        continue;
                    }
                    if !exists {
                        if !fix_refs {
                            continue;
                        }
                        remove_reference(&n, &attr, &value, &mut actions);
                    } else {
                        n.set(&attr, &uri.path);
                        actions.push(format!("Removed broken fragment: {value}"));
                    }
                    changed = true;
                    continue;
                }
                if !exists {
                    if fix_refs {
                        remove_reference(&n, &attr, &value, &mut actions);
                        changed = true;
                    }
                    continue;
                }
                if missing_entries
                    && !items
                        .iter()
                        .any(|i| archive::join(dir, &i.get("href")).is_ok_and(|p| p == target))
                    && let Some(media) = mime_guess::from_path(&target).first_raw()
                {
                    let stem = Path::new(&target)
                        .file_stem()
                        .unwrap_or_default()
                        .to_string_lossy();
                    let base = format!(
                        "optimizer-{}",
                        stem.chars()
                            .map(|c| if c.is_alphanumeric() { c } else { '-' })
                            .collect::<String>()
                            .trim_matches('-')
                    );
                    let mut id = base.clone();
                    let mut index = 2;
                    while ids.contains(&id) {
                        id = format!("{base}-{index}");
                        index += 1;
                    }
                    let href = archive::relative(package_path, &target);
                    let item = manifest.like("item");
                    item.set("id", &id);
                    item.set("href", &href);
                    item.set("media-type", media);
                    manifest.append(item.clone());
                    items.push(item);
                    ids.insert(id);
                    actions.push(format!(
                        "Added manifest item for referenced resource: {href}"
                    ));
                    changed = true;
                }
            }
        }
        if changed {
            xml.write(&work.join(path))?;
        }
    }
    Ok(actions)
}

fn repair_metadata(package: &Node, findings: &[crate::epubcheck::Finding]) -> Vec<String> {
    let mut actions = Vec::new();
    let codes: BTreeSet<_> = findings.iter().map(|f| f.code.to_uppercase()).collect();
    let Some(metadata) = package.child("metadata") else {
        return actions;
    };
    if codes.contains("OPF-030") {
        let identifiers: Vec<_> = metadata
            .children()
            .into_iter()
            .filter(|n| n.name() == "identifier" && !n.content().trim().is_empty())
            .collect();
        if let Some(chosen) = identifiers
            .iter()
            .find(|n| !n.get("id").is_empty())
            .or(identifiers.first())
        {
            let mut id = chosen.get("id");
            if id.is_empty() {
                id = "bookid".into();
                let mut index = 2;
                while metadata.children().iter().any(|n| n.get("id") == id) {
                    id = format!("bookid-{index}");
                    index += 1;
                }
                chosen.set("id", &id);
                actions.push(format!("Assigned stable id to dc:identifier: {id}"));
            }
            package.set("unique-identifier", &id);
            actions.push(format!(
                "Pointed package unique-identifier to dc:identifier: {id}"
            ));
        }
    }
    if codes.contains("OPF-028") {
        let re = regex::Regex::new(r#"(?i)prefix:\s*"([A-Za-z][\w-]*)""#).unwrap();
        for f in findings
            .iter()
            .filter(|f| f.code.to_uppercase() == "OPF-028")
        {
            if let Some(c) = re.captures(&f.message) {
                let prefix = &c[1];
                let prefixes = package.get("prefix");
                if prefix.eq_ignore_ascii_case("calibre")
                    && !prefixes
                        .split_whitespace()
                        .step_by(2)
                        .any(|p| p.strip_suffix(':') == Some(prefix))
                {
                    package.set(
                        "prefix",
                        format!("{prefixes} calibre: https://calibre-ebook.com").trim(),
                    );
                    actions.push("Declared known calibre metadata vocabulary".into());
                }
            }
        }
    }
    if codes.contains("RSC-005") {
        let re = regex::Regex::new(r#"Property "([^"]+)" must refine"#).unwrap();
        let invalid: BTreeSet<_> = findings
            .iter()
            .filter_map(|f| re.captures(&f.message).map(|c| c[1].to_string()))
            .collect();
        for n in metadata
            .children()
            .iter()
            .filter(|n| n.name() == "meta" && n.has("refines"))
        {
            let reference = n.get("refines");
            let property = n.get("property");
            let target = metadata.children().into_iter().find(|n| {
                reference
                    .strip_prefix('#')
                    .is_some_and(|r| n.get("id") == r)
            });
            let allowed = match property.as_str() {
                "role" => "creator contributor publisher",
                "file-as" => "creator contributor",
                "title-type" => "title",
                _ => "",
            };
            if property.is_empty()
                || target.is_none()
                || (invalid.contains(&property)
                    && !allowed.is_empty()
                    && target
                        .as_ref()
                        .is_some_and(|t| !crate::roles::one(&t.name(), allowed)))
            {
                n.remove_with_tail();
                actions.push("Removed invalid metadata refinement".into());
            }
        }
    }
    actions
}
fn remove_reference(n: &Node, attr: &str, value: &str, actions: &mut Vec<String>) {
    let local = n.name();
    let message = if matches!(local.as_str(), "a" | "area") && attr == "href" {
        n.del(attr);
        "Removed broken link target (text preserved)"
    } else if local == "img" {
        let alt = n.get("alt").trim().to_string();
        if !alt.is_empty() {
            let span = n.like("span");
            span.set_text(&alt);
            n.before(span);
            n.detach();
            "Replaced broken image with alt text"
        } else {
            n.remove_with_tail();
            "Removed broken image without alt text"
        }
    } else if matches!(local.as_str(), "object" | "audio" | "video") {
        n.del(attr);
        "Removed broken resource attribute; fallback preserved"
    } else if matches!(local.as_str(), "link" | "script" | "image" | "source") {
        n.remove_with_tail();
        "Removed broken resource element"
    } else {
        n.del(attr);
        "Removed broken reference"
    };
    actions.push(format!("{message}: {value}"));
}
